### PR TITLE
Add FQDN capability

### PR DESCRIPTION
**Pull request by ton31337**
It's not kinda RFC (draft), but it's implemented and used in various
other open-source software like FRRouting, Bird, ExaBGP.

It's very handy when dealing with lots of peers.

Exampe between GoBGP and FRRouting:
```
% ./cmd/gobgp/gobgp neighbor 192.168.10.123 | grep -A4 fqdn:
    fqdn:       advertised and received
      Local:
         name: donatas-pc, domain:
      Remote:
         name: exit1-debian-9, domain:
```

```
% vtysh -c 'show bgp neighbors 192.168.10.17 json' | jq .'"192.168.10.17".neighborCapabilities.hostName'
{
  "advHostName": "exit1-debian-9",
  "advDomainName": "n/a",
  "rcvHostName": "donatas-pc",
  "rcvDomainName": "n/a"
}
```

Signed-off-by: Donatas Abraitis <donatas.abraitis@gmail.com>

### CODE REVIEW

**Comment by fujita** [comment]
Looks good feature. Once the pull request passes the CI, I'm happy to push this feature.

**Comment by ton31337** [comment]
Thanks, @fujita. Can I run all these CI tests locally?

**Comment by fujita** [comment]
https://github.com/osrg/gobgp/blob/master/test/scenario_test/README.md

The static check isn't included. Please check out
https://github.com/osrg/gobgp/blob/master/.github/workflows/ci.yml#L33

**Comment by ton31337** [comment]
:+1:

**Comment by ton31337** [comment]
@fujita I've just ran failing tests locally and I can't reproduce that (for example `evpn` test):

```
gobgp/. [feature/FQDN_capability] () [10:36] donatas-pc donatas /home/donatas/projects/gobgp
% sudo -E PYTHONPATH=test python3 test/scenario_test/evpn_test.py --gobgp-image gobgp -x -s
[localhost] local: which docker 2>&1 > /dev/null ; echo $?
[localhost] local: if [ -e /tmp/gobgp//g1 ]; then rm -rf /tmp/gobgp//g1; fi
[localhost] local: mkdir -p /tmp/gobgp//g1
[localhost] local: chmod 777 /tmp/gobgp//g1
[localhost] local: docker ps -a | awk 'NR > 1 {print $NF}'
[localhost] local: docker rm -f g1
[localhost] local: if [ -e /tmp/gobgp//g2 ]; then rm -rf /tmp/gobgp//g2; fi
[localhost] local: mkdir -p /tmp/gobgp//g2
[localhost] local: chmod 777 /tmp/gobgp//g2
[localhost] local: docker ps -a | awk 'NR > 1 {print $NF}'
[localhost] local: docker rm -f g2
[g1's new gobgpd.conf]
    neighbors = []
    
    [defined-sets]
    
    [global.config]
    as = 65000
    router-id = "192.168.0.1"
    
    [global.route-selection-options.config]
    external-compare-router-id = true
    
    [global.use-multiple-paths.config]
    enabled = false
[localhost] local: docker run --privileged=true -v /tmp/gobgp//g1:/root/shared_volume -v /tmp/gobgp//g1/quagga:/etc/quagga --name g1 -l gobgp-test -id gobgp
[localhost] local: docker exec  g1 ip li set up dev lo
[localhost] local: docker exec  g1 ip a show dev eth0
[localhost] local: echo "#!/bin/sh
/go/bin/gobgpd -f /root/shared_volume/gobgpd.conf -l info -p  -t toml > /root/shared_volume/gobgpd.log 2>&1" > /tmp/gobgp//g1/start.sh
[localhost] local: chmod 755 /tmp/gobgp//g1/start.sh
[localhost] local: docker exec -d g1 /root/shared_volume/start.sh
[localhost] local: docker exec  g1 gobgp global > /dev/null 2>&1; echo $?
[g2's new gobgpd.conf]
    neighbors = []
    
    [defined-sets]
    
    [global.config]
    as = 65000
    router-id = "192.168.0.2"
    
    [global.route-selection-options.config]
    external-compare-router-id = true
    
    [global.use-multiple-paths.config]
    enabled = false
[localhost] local: docker run --privileged=true -v /tmp/gobgp//g2:/root/shared_volume -v /tmp/gobgp//g2/quagga:/etc/quagga --name g2 -l gobgp-test -id gobgp
[localhost] local: docker exec  g2 ip li set up dev lo
[localhost] local: docker exec  g2 ip a show dev eth0
[localhost] local: echo "#!/bin/sh
/go/bin/gobgpd -f /root/shared_volume/gobgpd.conf -l info -p  -t toml > /root/shared_volume/gobgpd.log 2>&1" > /tmp/gobgp//g2/start.sh
[localhost] local: chmod 755 /tmp/gobgp//g2/start.sh
[localhost] local: docker exec -d g2 /root/shared_volume/start.sh
[localhost] local: docker exec  g2 gobgp global > /dev/null 2>&1; echo $?
[localhost] local: docker exec  g1 gobgp vrf add vrf1 rd 10:10 rt both 10:10
[localhost] local: docker exec  g2 gobgp vrf add vrf1 rd 10:10 rt both 10:10
[g1's new gobgpd.conf]
    [[neighbors]]
    
    [[neighbors.afi-safis]]
    
    [neighbors.afi-safis.config]
    afi-safi-name = "ipv4-unicast"
    [[neighbors.afi-safis]]
    
    [neighbors.afi-safis.config]
    afi-safi-name = "l3vpn-ipv4-unicast"
    [[neighbors.afi-safis]]
    
    [neighbors.afi-safis.config]
    afi-safi-name = "l3vpn-ipv6-unicast"
    [[neighbors.afi-safis]]
    
    [neighbors.afi-safis.config]
    afi-safi-name = "l2vpn-evpn"
    [[neighbors.afi-safis]]
    
    [neighbors.afi-safis.config]
    afi-safi-name = "rtc"
    [neighbors.afi-safis.route-target-membership.config]
    deferral-time = 10
    [neighbors.config]
    neighbor-address = "172.17.0.7"
    peer-as = 65000
    auth-password = "evpn"
    vrf = ""
    [neighbors.timers.config]
    connect-retry = 10
    
    [defined-sets]
    
    [global.config]
    as = 65000
    router-id = "192.168.0.1"
    
    [global.route-selection-options.config]
    external-compare-router-id = true
    
    [global.use-multiple-paths.config]
    enabled = false
[localhost] local: docker exec  g1 pkill -SIGHUP gobgpd
[localhost] local: docker exec  g1 gobgp global > /dev/null 2>&1; echo $?
[g2's new gobgpd.conf]
    [[neighbors]]
    
    [[neighbors.afi-safis]]
    
    [neighbors.afi-safis.config]
    afi-safi-name = "ipv4-unicast"
    [[neighbors.afi-safis]]
    
    [neighbors.afi-safis.config]
    afi-safi-name = "l3vpn-ipv4-unicast"
    [[neighbors.afi-safis]]
    
    [neighbors.afi-safis.config]
    afi-safi-name = "l3vpn-ipv6-unicast"
    [[neighbors.afi-safis]]
    
    [neighbors.afi-safis.config]
    afi-safi-name = "l2vpn-evpn"
    [[neighbors.afi-safis]]
    
    [neighbors.afi-safis.config]
    afi-safi-name = "rtc"
    [neighbors.afi-safis.route-target-membership.config]
    deferral-time = 10
    [neighbors.config]
    neighbor-address = "172.17.0.2"
    peer-as = 65000
    auth-password = "evpn"
    vrf = ""
    [neighbors.timers.config]
    connect-retry = 10
    
    [defined-sets]
    
    [global.config]
    as = 65000
    router-id = "192.168.0.2"
    
    [global.route-selection-options.config]
    external-compare-router-id = true
    
    [global.use-multiple-paths.config]
    enabled = false
[localhost] local: docker exec  g2 pkill -SIGHUP gobgpd
[localhost] local: docker exec  g2 gobgp global > /dev/null 2>&1; echo $?
[localhost] local: docker exec  g1 gobgp -j neighbor 172.17.0.7
192.168.0.1's peer 192.168.0.2 state: active
[localhost] local: docker exec  g1 gobgp -j neighbor 172.17.0.7
192.168.0.1's peer 192.168.0.2 state: active
[localhost] local: docker exec  g1 gobgp -j neighbor 172.17.0.7
192.168.0.1's peer 192.168.0.2 state: active
[localhost] local: docker exec  g1 gobgp -j neighbor 172.17.0.7
192.168.0.1's peer 192.168.0.2 state: active
[localhost] local: docker exec  g1 gobgp -j neighbor 172.17.0.7
192.168.0.1's peer 192.168.0.2 state: established
.[localhost] local: docker exec  g1 gobgp global rib add -a evpn macadv 11:22:33:44:55:66 10.0.0.1 esi AS 1 1 1 etag 1000 label 1000 rd 10:10 rt 10:10
[localhost] local: docker exec  g1 gobgp -j global rib  -a evpn
[localhost] local: docker exec  g2 gobgp -j global rib  -a evpn
.[localhost] local: docker exec  g2 gobgp global rib add -a evpn macadv 11:22:33:44:55:66 10.0.0.1 esi AS 2 1 1 etag 1000 label 1000 rd 10:20 rt 10:10
[localhost] local: docker exec  g1 gobgp -j global rib  -a evpn
.[localhost] local: docker exec  g1 gobgp global rib add -a evpn macadv 11:22:33:44:55:66 10.0.0.1 esi AS 3 1 1 etag 1000 label 1000 rd 10:20 rt 10:10
[localhost] local: docker exec  g2 gobgp -j global rib  -a evpn
.
----------------------------------------------------------------------
Ran 4 tests in 17.236s

OK
gobgp/. [feature/FQDN_capability] () [10:36] donatas-pc donatas /home/donatas/projects/gobgp
```

**Comment by fujita** [comment]
You build docker image from your branch before running the evpn test? I can reproduce the error locally.

File: api/capability.proto [hunk 1]
```diff
@@ -94,7 +94,14 @@ message LongLivedGracefulRestartCapability {
 message RouteRefreshCiscoCapability {
 }
 
+message FQDNCapability {
+    uint32 host_name_len = 1;
+    bytes host_name = 2;
```
Comment (fujita) [inline api/capability.proto:99]: why not string?

File: api/capability.proto [hunk 2]
```diff
@@ -94,7 +94,14 @@ message LongLivedGracefulRestartCapability {
...
+    uint32 domain_name_len = 3;
+    bytes domain_name = 4;
+}
```
Comment (fujita) [inline api/capability.proto:102]: ditto.

File: pkg/packet/bgp/bgp.go [hunk 3]
```diff
@@ -916,6 +918,65 @@ func NewCapLongLivedGracefulRestart(tuples []*CapLongLivedGracefulRestartTuple)
 	}
 }
 
+type CapFQDN struct {
+	DefaultParameterCapability
+	HostNameLen   uint8
+	HostName      []byte
```
Comment (fujita) [inline pkg/packet/bgp/bgp.go:924]: string?

File: pkg/packet/bgp/bgp.go [hunk 4]
```diff
@@ -916,6 +918,65 @@ func NewCapLongLivedGracefulRestart(tuples []*CapLongLivedGracefulRestartTuple)
...
+	DomainNameLen uint8
+	DomainName    []byte
+}
```
Comment (fujita) [inline pkg/packet/bgp/bgp.go:927]: ditto

File: api/capability.proto (hunk 1 above)
Comment (ton31337) [reply api/capability.proto:99]: No real reason, will change to string.

**Comment by ton31337** [comment]
Seems good now.

**Comment by ton31337** [comment]
@fujita all good?

**Comment by fujita** [comment]
looks ok, thanks!

**Pull request merged**